### PR TITLE
Busy wait utils in dist

### PR DESCRIPTION
**Pull request by dakinggg**
# What does this PR do?
This PR adds additional utils for busy waiting on a node using a signal file lock. The main difference between the added utils and the current approach used is that the signal file has a randomly generated identified appended to it, to better support use cases that include multiple runs using a shared file system that may interfere with each other. Future PRs will replace the signal file uses in Composer (and after release Foundry) with these utils, but just want to start with the util implementation to keep PRs small.

# What issue(s) does this change relate to?
Related to https://github.com/mosaicml/llm-foundry/pull/1253

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by mvpatel2000 — COMMENTED** [review]
can you update calls in composer to use these helper fns?

File: composer/utils/dist.py [hunk 1]
```diff
@@ -627,6 +629,76 @@ def get_sampler(
...
+    Returns:
+        str: The name of the file that will be created to signal the end of a node's training.
+    """
+    if rng is None:
+        rng = random.Random()
+
+    random_string = ''.join(rng.choices(string.ascii_letters + string.digits, k=6))
+    file_name_list = [f'._signal_file_{random_string}']
+    dist.broadcast_object_list(file_name_list, src=0)
+    return file_name_list[0]
```
Comment (mvpatel2000) [inline composer/utils/dist.py:647]: this returns the global rank 0? is that intnded

Comment (mvpatel2000) [reply composer/utils/dist.py:647]: iirc we want different names per node to avoid shared file system cases

File: composer/utils/dist.py [hunk 2]
```diff
@@ -627,6 +629,76 @@ def get_sampler(
...
+def write_signal_file(signal_file_name: str, dir_path: Optional[str] = None) -> str:
+    """Writes a signal file to the specified directory.
+
+    This function creates a signal file in the specified directory. If the directory does
+    Note: Only local rank zero writes the signal file. All other ranks are expected to wait for the signal file.
+
+    Args:
+        signal_file_name (str): The name of the signal file.
+        dir_path (str, optional): The directory in which to create the signal file. If ``None``,
+            the current working directory will be used.
```
Comment (mvpatel2000) [inline composer/utils/dist.py:659]: should specify its an absolute dir path

File: composer/utils/dist.py (hunk 1 above)
Comment (dakinggg) [reply composer/utils/dist.py:647]: ah thank you, i waffled on this but couldn't remember if there was a reason why they were supposed to be node specific. will adjust.

Comment (dakinggg) [reply composer/utils/dist.py:648]: fixed

File: composer/utils/dist.py (hunk 2 above)
Comment (dakinggg) [reply composer/utils/dist.py]: done

File: composer/utils/dist.py [hunk 3]
```diff
@@ -627,6 +629,77 @@ def get_sampler(
...
+
+    # All ranks yield execution to allow local rank zero to run the code it needs to
+    yield
+
+    # Local rank zero writes the signal file, all other rank just get the expected path
+    signal_file_path = write_signal_file(signal_file_name=signal_file_name, dir_path=dir_path)
+
+    # Wait for the signal file to be created by local rank zero
+    with local_rank_zero_download_and_wait(signal_file_path):
+        # Sync up all ranks
```
Comment (mvpatel2000) [inline composer/utils/dist.py:695]:
```suggestion
        # Sync all ranks across nodes as busy wait only is within node
```

**Review by mvpatel2000 — APPROVED** [review]
LGTM. Minor comment

**Pull request merged**